src/nodes.coffee

```diff
@@ -273,6 +278,15 @@ exports.Base = class Base
   ast: (o, level) ->
     o = Object.assign {}, o
     o.level = level if level?
+    @makeReturn null, yes if @isStatement(o) and o.level isnt LEVEL_TOP and o.scope?
+
+    ast = @astNode o
+    return ast unless ast?
+    # Mark AST nodes that correspond to expressions that (implicitly) return.
+    ast.returns = yes if @canBeReturned
+    ast
```

Comment:
This is only for organization/clarity; what do you think of this?

Something like:

```coffee
  ast: (o, level) ->
    o = astInitialize o, level
    astNode o # Calls `astReturn` to return final assembled object.

  # The “constructor” for AST nodes, this logic must be run as part of creating any
  # AST node.
  astInitialize: (o, level) ->
    o = Object.assign {}, o
    o.level = level if level?
    @makeReturn null, yes if @isStatement(o) and o.level isnt LEVEL_TOP and o.scope?
    o

  # If an AST node needs any unusual logic that doesn’t fit within the component
  # methods `astType` or `astProperties`, such as using a different AST class
  # entirely, that can go in `astNode`. If you override this, make sure to call
  # `@astReturn o` at the end of your method.
  astNode: (o) -> astReturn o

  # Assemble the serializable JSON-like AST object to be returned.
  astReturn: (o) ->
    # Mark AST nodes that correspond to expressions that (implicitly) return.
    returns = if @canBeReturned then {returns: yes} else {}

    # Every abstract syntax tree node object has four categories of properties:
    # - type, stored in the `type` field and a string like `NumberLiteral`.
    # - location data, stored in the `loc`, `start`, `end` and `range` fields.
    # - properties specific to this node, like `parsedValue`.
    # - properties that are themselves child nodes, like `body`.
    # These fields are all intermixed in the Babel spec; `type` and `start` and
    # `parsedValue` are all top level fields in the AST node object. We have
    # separate methods for returning each category, that we merge together here.
    Object.assign {}, {type: @astType(o)}, returns, @astProperties(o), @astLocationData()
```

Reply:
Ok I updated to use `astInitialize()`

The issue with nesting `astReturn()` inside `astNode()` is that then node classes that override `astNode()` wouldn't get the `returns` field added automatically - I think the adding of `returns` as the last step of assembling the AST node should remain a responsibility of `ast()` rather than `astNode()` so that it can stay centralized in a single place rather than being something that then every node class that overrides `astNode()` would have to separately know how to do

But so I updated to make the adding of `returns` a little more explicit by encapsulating in an `astAddReturns()` method (that gets applied as the last step of `Base::ast()`)